[port] Few Minor per-utxo assert-semantics re-adds and tweak

sickpig opened the pull request:
This is port of bitcoin/bitcoin/pull/10537

TheBlueMatt committed:
Remove useless mapNextTx lookup in CTxMemPool::TrimToSize.

Prior to per-utxo CCoins, we checked that no other in-mempool tx
spent any of the given transaction's outputs, as we don't want to
uncache that entire tx in such a case. However, we now are checking
only that there exists no other mempool spends of the same output,
which should clearly be impossible after we removed the transaction
which was spending said output (barring massive mempool
inconsistency).

Thanks to @sdaftuar for the suggestion.

TheBlueMatt committed:
Return a bool in SpendCoin to restore pre-per-utxo assert semantics

Since its free to do so, assert that Spends succeeded when we expect
them to.

TheBlueMatt committed:
Be much more agressive in AccessCoin docs.

While the current implementation is pretty free, there is a lot
of possibility for this to blow up in our face with future changes,
especially as the backing map gets tweaked.

Greg-Griffith commented on src/consensus/tx_verify.cpp:
```diff
@@ -149,6 +149,8 @@ unsigned int GetP2SHSigOpCount(const CTransactionRef &tx, const CCoinsViewCache
         {
             CoinAccessor coin(inputs, tx->vin[i].prevout);
+            assert(!coin.IsSpent());
+            const CTxOut &prevout = coin.out;
```
the assert here should be changed to return some error instead of asserting because there is no indication that the tx needs to be verified before running this function. 

line 153 prevout variable is unused

TheBlueMatt committed:
Restore some assert semantics in sigop cost calculations

There are some similar asserts which are left removed in policy
and ATMP (policy code being broken isn't a huge deal, but if we
fail to verify some consensus rules, we should most definitely
crash).

sickpig committed:
Update commit for GetP2SHSigOpCount

This function is supposed to be used once we are sure that inputs
have already been validated.

gandrewstone commented on src/coins.h:
```diff
@@ -301,8 +301,13 @@ class CCoinsViewCache : public CCoinsViewBacked
+     * While the current implementation allows for modifications to the contents
+     * of the cache while holding the reference, this behavior should not be relied
+     * on! To be safe, best to not hold the returned reference through any other
+     * calls to this cache.
```
Due to the multi-threaded nature of the BU code, its dangerous to use this at all, which is why it lack of locking is indicated by prepending the function with an _.  In BU, the "CoinAccessor" wrapper should be used since it properly handles locking over the duration of your use of the coin.  (look at your change in tx_verify.cpp to see a use a CoinAccessor)  We should change this comment to reflect BU's greater strictness requirement.

sickpig replied:
fair enough, will do.

sickpig replied:
done

sickpig committed:
Update description comment for _AccessCoin

[skip ci]

sickpig commented:
@gandrewstone I addressed your comment and change code comment accordingly.
